PythonMutator: add diagnostics

kanterov opened the pull request:
## Changes
Allow PyDABs to report `dyn.Diagnostics` by writing to `diagnostics.json` supplied as an argument, similar to `input.json` and `output.json`

Such errors are not yet properly printed in `databricks bundle validate`, which will be fixed in a follow-up PR.

## Tests
Unit tests

pietern commented on bundle/config/mutator/python/python_diagnostics.go:
```diff
@@ -0,0 +1,129 @@
+
+		parsedLine := pythonDiagnostic{}
+
+		err := json.Unmarshal([]byte(line), &parsedLine)
```
You can use https://pkg.go.dev/encoding/json#NewDecoder and not deal with the newlines.

It works with line-delimited JSON objects out of the box.

kanterov replied:
Thanks! This is much easier now

kanterov commented:
@pietern I've addressed your comments. Please take a look again.

kanterov commented on bundle/config/mutator/python/python_mutator.go:
```diff
@@ -103,9 +106,10 @@ func (m *pythonMutator) Apply(ctx context.Context, b *bundle.Bundle) diag.Diagno
-		rightRoot, err := m.runPythonMutator(ctx, cacheDir, b.RootPath, pythonPath, leftRoot)
-		if err != nil {
-			return dyn.InvalidValue, err
+		rightRoot, diagnostics := m.runPythonMutator(ctx, cacheDir, b.RootPath, pythonPath, leftRoot)
```
Renaming to `diag` will shadow package import for `diag`, I will use `diags` instead.

It doesn't directly use `mutateDiagnostics` because we want both to declare a new variable, and mutate an existing one, and we never want to shadow `mutateDiagnostics`. This can be very error-prone, so we make it very explicit in this case.

kanterov commented on bundle/config/mutator/python/python_mutator.go:
```diff
@@ -152,42 +161,71 @@ func (m *pythonMutator) runPythonMutator(ctx context.Context, cacheDir string, r
 		"--output",
 		outputPath,
+		"--diagnostics",
+		diagnosticsPath,
```
It doesn't have to be consistent because changes need to be backward and forward-compatible. If we make a breaking change, we raise the minimum required version.

pietern replied:
Where is the minimum required version stored/communicated?

kanterov replied:
We don't. For now, we can add a minimal required version of CLI as the configuration in the template, and document it.

We can also check in VerifyCliVersion, but right now, it's executed before we run the Python mutator if we want to append a different minimum required version there. Another difficulty is that this configuration parameter might already be used. 

Another way is to pass the current CLI version into Python code and check it there. We can make it a part of the effective bundle configuration.

kanterov commented on bundle/config/mutator/python/python_mutator.go:
```diff
@@ -152,42 +161,71 @@ func (m *pythonMutator) runPythonMutator(ctx context.Context, cacheDir string, r
+
+	// if diagnostics file exists, it gives the most descriptive errors
+	// if there is any error, we treat it as fatal error, and stop processing
+	if pythonDiagnostics.HasError() {
```
`pythonDiagnosticsErr` should always be nil if `pythonDiagnostics` has an error

The idea behind `pythonDiagnostics` is that we should never show a "process exit" error if it contains an explanation. This is how PyDABs explain what went wrong. There are 3 errors possible, and you can see the code below as the way for us to choose the most informative error that can explain what went wrong, in the order:

- PyDABs reported an error that we loaded/parsed succesfuly
- Process exited with non-zero code
- We failed to load/parse the diagnostics file, but the process exited successfully

pietern replied:
The error handling flow is not super straightforward. Could you include a few debug log lines to capture error conditions even if we don't act on them? When this goes wrong I'd like to be able to analyse why/how/etc.

kanterov replied:
Agree. I added logging just after each error is created

kanterov commented on bundle/config/mutator/python/python_diagnostics.go:
```diff
@@ -0,0 +1,121 @@
+	"github.com/databricks/cli/libs/dyn"
+)
+
+type pythonDiagnostic struct {
```
I did research on how other tools report warnings:
- go vet
- eslint
- pylint
- terraform validate

All k8s tools report the path as a string. I will change it to do the same.

kanterov replied:
Also note that this format is purely internal, and we wouldn't show this to users, occasionally, we might need to read it for debugging.

If we ever want CLI to report warnings in a structured format, it can use any format not necessarily the same as here.

pietern replied:
+1 -- I wouldn't want to couple them as they might need to evolve independently.

kanterov commented on bundle/config/mutator/python/python_mutator.go:
```diff
@@ -116,7 +120,11 @@ func (m *pythonMutator) Apply(ctx context.Context, b *bundle.Bundle) diag.Diagno
 	})
 
-	return diag.FromErr(err)
+	if len(mutateDiagnostics) != 0 {
```
We can't merge errors because they can create duplicates. I will refactor them to make it easier to follow.

pietern commented on bundle/config/mutator/python/python_diagnostics.go:
```diff
@@ -0,0 +1,98 @@
+
+		err := decoder.Decode(&parsedLine)
+
+		if err != nil {
```
Style: superfluous newline

kanterov replied:
Fixed, bad habits :)

pietern commented on bundle/config/mutator/python/python_mutator.go:
```diff
@@ -204,24 +248,38 @@ func (m *pythonMutator) runPythonMutator(ctx context.Context, cacheDir string, r
+//
+// It contains a list of warnings and errors that we should print to users.
+func loadDiagnosticsFile(path string) (diag.Diagnostics, error) {
+	file, err := os.Open(path)
```
Is it guaranteed to exist, even if there are no diagnostics?

kanterov replied:
Yes, I think it will be less error-prone like that, so it doesn't happen that integration suddenly breaks and stops to print warnings and errors

pietern replied:
Please capture this intent in the comment. The next person looking at this code could add a not-exists check not being aware of this intent.

pietern commented on bundle/config/mutator/python/python_mutator_test.go:
```diff
@@ -74,12 +74,14 @@ func TestPythonMutator_load(t *testing.T) {
+	diagnostics := bundle.Apply(ctx, b, mutator)
 
-	assert.NoError(t, diag.Error())
+	assert.NoError(t, diagnostics.Error())
```
Convention: `diags`

kanterov commented:
@shreyas-goenka @pietern addressed your comments, please take a look.

The remaining question is the CLI version check. I don't find it pressing if we put effort into templates and documentation. I think in the long run, we should serialize the current CLI version as a part of the JSON we pass. What do you think?

pietern reviewed (approved):
One minor last comment

kanterov commented:
@pietern done, I've added an elaborate comment and test. Can you please add it to the merge queue?

@pietern @shreyas-goenka thanks for taking the look
